src/controller_examples/zookeeper_controller/proof/liveness.rs

```diff
@@ -1192,17 +897,113 @@ proof fn lemma_true_leads_to_always_current_state_matches_zk_under_eventual_inva
         spec,
         true_pred(),
         lift_state(|s: ClusterState| !s.reconcile_state_contains(zk.object_ref())),
-        lift_state(at_init_step_with_no_pending_req(zk)),
-        lift_state(at_after_create_headless_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_client_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_admin_server_service_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_create_config_map_step_with_zk_and_pending_req_in_flight(zk)),
-        lift_state(at_after_get_stateful_set_step_with_zk_and_pending_req_in_flight(zk)),
+        lift_state(no_pending_req_at_zookeeper_step_with_zk(zk, ZookeeperReconcileStep::Init)),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateHeadlessService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateClientService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateAdminServerService, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterCreateConfigMap, zk, arbitrary())),
+        lift_state(pending_req_in_flight_at_zookeeper_step_with_zk(ZookeeperReconcileStep::AfterGetStatefulSet, zk, arbitrary())),
         lift_state(current_state_matches(zk)),
         always(lift_state(current_state_matches(zk)))
     );
 }
 
+proof fn lemma_from_pending_req_in_flight_at_some_step_to_next_step(
+    spec: TempPred<ClusterState>, zk: ZookeeperClusterView, step: ZookeeperReconcileStep, next_step: ZookeeperReconcileStep
+)
+    requires
+        spec.entails(always(lift_action(next::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>()))),
+        spec.entails(tla_forall(|i| controller_next::<ZookeeperClusterView, ZookeeperReconcileState, ZookeeperReconciler>().weak_fairness(i))),
+        spec.entails(tla_forall(|i| kubernetes_api_next().weak_fairness(i))),
+        spec.entails(always(lift_state(crash_disabled()))),
+        spec.entails(always(lift_state(controller_runtime_safety::every_in_flight_msg_has_unique_id()))),
+        spec.entails(always(lift_state(controller_runtime_safety::each_resp_matches_at_most_one_pending_req(zk.object_ref())))),
+        step != ZookeeperReconcileStep::Error, step != ZookeeperReconcileStep::Done,
+        next_step != ZookeeperReconcileStep::Init,
+        forall |zk_1: ZookeeperClusterView, resp_o: Option<APIResponse>|
+            #[trigger] reconcile_core(zk_1, resp_o, ZookeeperReconcileState{ reconcile_step: step }).0.reconcile_step == next_step
```

Comment:
I think this precondition wants to say no matter what, the next step is `next_step`. I am trying to understand why we have to forall zk_1 here, instead of just talking about zk.

Reply:
We can't infer from the `pre` state that the cr passed to `reconcile_core` is exactly `zk`. However, we may weaken the precondition for this lemma.